pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,426 @@
+				cluster.Status.ServiceAccountToken = serviceToken
+				return e.ClusterClient.Update(cluster)
+			}
+		}
+
+		if cluster.Status.ServiceAccountToken == "" {
+			cluster, err = e.generateAndSetServiceAccount(cluster)
+			if err != nil {
+				var statusErr error
+				if strings.Contains(err.Error(), fmt.Sprintf(dialer.WaitForAgentError, cluster.Name)) {
+					// In this case, the API endpoint is private and rancher is waiting for the import cluster command to be run.
+					cluster, statusErr = e.SetUnknown(cluster, apimgmtv3.ClusterConditionWaiting, "waiting for cluster agent to be deployed")
+					if statusErr == nil {
+						e.ClusterEnqueueAfter(cluster.Name, enqueueTime)
+					}
+					return cluster, statusErr
+				}
+				cluster, statusErr = e.SetFalse(cluster, apimgmtv3.ClusterConditionWaiting,
+					fmt.Sprintf("failed to communicate with cluster: %v", err))
+				if statusErr != nil {
+					return cluster, statusErr
+				}
+				return cluster, err
+			}
+		}
+
+		cluster, err = e.recordAppliedSpec(cluster)
+		if err != nil {
+			return cluster, err
+		}
+		return e.SetTrue(cluster, apimgmtv3.ClusterConditionUpdated, "")
+	case "updating":
+		cluster, err = e.SetTrue(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		if err != nil {
+			return cluster, err
+		}
+
+		e.ClusterEnqueueAfter(cluster.Name, enqueueTime)
+		if failureMessage == "" {
+			logrus.Infof("waiting for cluster AKS [%s] to update", cluster.Name)
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		}
+		logrus.Infof("waiting for cluster AKS [%s] update failure to be resolved", cluster.Name)
+		return e.SetFalse(cluster, apimgmtv3.ClusterConditionUpdated, failureMessage)
+	default:
+		if cluster.Spec.AKSConfig.Imported {
+			cluster, err = e.SetUnknown(cluster, apimgmtv3.ClusterConditionPending, "")
+			if err != nil {
+				return cluster, err
+			}
+			logrus.Infof("waiting for cluster import [%s] to start", cluster.Name)
+		} else {
+			logrus.Infof("waiting for cluster create [%s] to start", cluster.Name)
+		}
+
+		e.ClusterEnqueueAfter(cluster.Name, enqueueTime)
+		if failureMessage == "" {
+			if cluster.Spec.AKSConfig.Imported {
+				cluster, err = e.SetUnknown(cluster, apimgmtv3.ClusterConditionPending, "")
+				if err != nil {
+					return cluster, err
+				}
+				logrus.Infof("waiting for cluster import [%s] to start", cluster.Name)
+			} else {
+				logrus.Infof("waiting for cluster create [%s] to start", cluster.Name)
+			}
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		}
+		logrus.Infof("waiting for cluster AKS [%s] pre-create failure to be resolved", cluster.Name)
+		return e.SetFalse(cluster, apimgmtv3.ClusterConditionProvisioned, failureMessage)
+	}
+}
+
+func (e *aksOperatorController) setInitialUpstreamSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	logrus.Infof("setting initial upstreamSpec on cluster [%s]", cluster.Name)
+	cluster = cluster.DeepCopy()
```

Comment:
nit: Only perform this `DeepCopy` if we need to and move after error check below (L280)